sklearn/utils/testing.py

```diff
@@ -422,6 +463,166 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
+    function(s) are similar, a comparison of the attributes of the models
+    that end with ``_`` is done to ascertain the similarity of the model.
+
+    If the models are different an AssertionError with the given error message
+    is raised.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        Input data, for the fitted models, used for comparing them.
+
+    estimator1 : An estimator object.
+        The first fitted model to be compared.
+
+    estimator2 : An estimator object.
+        The second fitted model to be compared.
+
+    msg : str
+        The error message to be used while raising the AssertionError if the
+        models are similar.
+
+    Notes
+    -----
+    This check is not exhaustive since all attributes of the model are assumed
+    to end with ``_``. If that is not the case, it could lead to false
+    positives.
+    """
+    _assert_same_model_method('predict', X, estimator1, estimator2, msg)
+    _assert_same_model_method('transform', X, estimator1, estimator2, msg)
+    _assert_same_model_method('decision_function',
+                              X, estimator1, estimator2, msg)
+    _assert_same_model_method('predict_proba', X, estimator1, estimator2, msg)
+    assert_fitted_attributes_almost_equal(estimator1, estimator2)
+
+
+def assert_not_same_model(X, estimator1, estimator2, msg=None):
+    """Helper function to check if the models are different.
+
+    The check is done by comparing the outputs of the methods ``predict``,
+    ``transform``, ``decision_function`` and the ``predict_proba``, provided
+    they exist in both the models. If any of those methods do not exist in
+    one model alone, the models are considered different.
+
+    If the outputs from both the models for each of the available, above listed
+    function(s) are similar, a comparison of the attributes of the models
+    that end with ``_`` is done to ascertain the similarity of the model.
+
+    If the models are similar an AssertionError with the given error message
+    is raised.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        Input data, for the fitted models, used for comparing them.
+
+    estimator1 : An estimator object.
+        The first fitted model to be compared.
+
+    estimator2 : An estimator object.
+        The second fitted model to be compared.
+
+    msg : str
+        The error message to be used while raising the AssertionError if the
+        models are similar.
+
+    Notes
+    -----
+    This check is not exhaustive since all attributes of the model are assumed
+    to end with ``_``. If that is not the case, it could lead to false
+    negatives.
+    """
+    try:
+        assert_same_model(X, estimator1, estimator2)
+    except AssertionError:
+        pass
+    else:
+        raise AssertionError(msg)
+
+
+def assert_fitted_attributes_almost_equal(estimator1, estimator2, msg=None):
```

Comment:
I find it hard to follow why this doesn't have an underscore but other asserters introduced here do.